desktop_version/src/DiscordNetwork.c

```diff
@@ -0,0 +1,183 @@
+FUNC_LIST
+#undef FOREACH_FUNC
+
+bool DISCORD_REQUIRE(int x)
+{
+    if (!discordDetected && discordPostInit)
+    {
+        return false;
+    }
+    if (x != DiscordResult_Ok)
+    {
+        return false;
+    }
+    return true;
+}
+
+
+
+static void ClearPointers(void)
+{
+    SDL_UnloadObject(libHandle);
+    libHandle = NULL;
+#define FOREACH_FUNC(rettype, name, params) name = NULL;
+    FUNC_LIST
+#undef FOREACH_FUNC
+}
+
+void DISCORD_shutdown(void)
+{
+    if (libHandle != NULL)
+    {
+        app.core->destroy(app.core);
+        ClearPointers();
+    }
+}
+
+int32_t DISCORD_init(void)
+{
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__HAIKU__) || defined(__DragonFly__)
+    return 0;
+#endif
+    libHandle = SDL_LoadObject(DISCORD_LIBRARY);
+    if (libHandle == NULL)
+    {
+        vlog_info(DISCORD_LIBRARY " not found!");
+        vlog_debug("Can't load object %s : %s\n", DISCORD_LIBRARY, SDL_GetError());
+        return 0;
+    }
+
+#define FOREACH_FUNC(rettype, name, params) \
+    name = (rettype (*) params) (intptr_t) SDL_LoadFunction(libHandle, #name); \
+    if (!name) \
+    { \
+        vlog_error(DISCORD_LIBRARY " symbol " #name " not found!"); \
+        ClearPointers(); \
+        return 0; \
+    }
+    FUNC_LIST
+#undef FOREACH_FUNC
+    SDL_memset(&app, 0, sizeof(app));
+
+    struct DiscordCreateParams params;
+    params.client_id = DISCORD_CLIENT_ID;
+    params.flags = DiscordCreateFlags_NoRequireDiscord;
+
+    if (!DISCORD_REQUIRE(DiscordCreate(DISCORD_VERSION, &params, &app.core)))
+    { // Discord's API couldn't initialise, so just ignore it
+        discordPostInit = true; // even if it fails, set post init to true.
+        vlog_debug("Discord API failed to initialise!");
+        discordDetected = false;
+        return 0;
+    }
+
+    if (app.core == NULL)
+    {
+        discordPostInit = true;
+        vlog_debug("app.core == null. DiscordCreate failed with a positive result?\nCheck DISCORD_REQUIRE  for bugs.");
+        discordDetected = false;
+        return 0;
+    }
+    discordPostInit = true;
+
+    // Placeholder, is this really nesaccary
+    SDL_strlcpy(activity.assets.large_image, "vvvvvv", sizeof(activity.assets.large_image));
+    SDL_strlcpy(activity.assets.large_text, "Outside Dimension VVVVVV", sizeof(activity.assets.large_text));
+
+    app.activityMan = app.core->get_activity_manager(app.core);
+    app.activityMan->update_activity(app.activityMan, &activity, NULL, NULL);
+
+
+    discordDetected = true;
+    return 1;
+}
+
+void DISCORD_update(const char* area, const char* roomname)
+{
+    if (libHandle == NULL)
+    {
+        // no discord or just shutdown
+        return;
+    }
+    if (app.core == NULL || !discordDetected)
+    {
+        // No Discord
+        return;
+    }
+    if (!DISCORD_REQUIRE(app.core->run_callbacks(app.core)))
+    {
+        // Something  or other is wrong, but do we care?
```

Comment:
Probably not, but Discord support might...? Not sure what their policy is on reporting internal errors